### ISSUE TITLE
Reference specifications and implementations

### ISSUE THREAD

**Issue by shadowspawn** [pkgjs/parseargs#76]
I was thinking about what to consult for command-line parsing behaviours.

The two specifications I am aware of are:
- POSIX which supports short options only, options must come before positional
  - https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html#tag_12_01
  - https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_xbd_chap12.html
- GNU adds long options, adds mixed order options and positionals, adds options with optional arguments
  - https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html
  - https://www.gnu.org/prep/standards/html_node/Command_002dLine-Interfaces.html

With reference implementations  for C programs:
- [getopt](https://pubs.opengroup.org/onlinepubs/9699919799/functions/getopt.html) for POSIX
- [getopt_long](https://www.gnu.org/software/libc/manual/html_node/Getopt.html) and [argp_parse](https://www.gnu.org/software/libc/manual/html_node/Argp.html) for GNU

-----

In Javascript world, we can look at npm dependencies as one measure of usage. Packages with over 1000 dependents:

| dependants | package |
| --- | --- |
| 61k | commander |
|28k | yargs (uses yargs-parser as engine) |
| 19k | minimist |
| 6k | meow (uses yargs-parser as engine) |
| 4k | @oclif/command |
| 2k | command-line-args |
| 2k | yargs-parser |
| 2k | argparse |
| 1k | nopt |

- minimist and yargs share the same roots
- meow uses yargs-parser as engine
- minimist is in maintenance mode.
- assuming yargs and yargs-parser are distinct for counts

For a second source for usage, this just came out: [Census II of Free and Open Source Software — Application Libraries](https://www.linuxfoundation.org/tools/census-ii-of-free-and-open-source-software--application-libraries/)

In the top 500 npm packages as direct dependencies:

| z-score | package | rank |
|--- | --- | --- |
| 2.2 | commander | 23 |
| 2.1 | minimist | 28 |
| 1.1 |  yargs | 73 |
| 1.1 | yargs-parser | 75 |
| 1.1 | @radic/yargs (?) | 77 |
| 0.3 | nopt | 212 |
| 0.05 | argparse | 344 |

Across both measures, commander and yargs and minimist are the big 3 covering the majority of direct dependencies.

**Comment by shadowspawn** [comment]
For interest, [minimist](https://www.npmjs.com/package/minimist) is currently unmaintained. The GitHub source repo has been deleted.

**Comment by ljharb** [comment]
Yes, sadly substack deleted all their github repos. If @emilbayes is willing to `npm owner add` me, i'd be happy to reconstruct the repo and continue to maintain it.

**Comment by shadowspawn** [comment]
This fork has been active this year: https://github.com/meszaros-lajos-gyorgy/minimist-lite

**Comment by ljharb** [comment]
I've got the original git history up to v1.2.0 in 2015; a link to a repo that has commits through March 2022 would be appreciated :-)

**Comment by shadowspawn** [comment]
To be clear, I was pointing out an active fork by people who have already shown an interest and investment in maintaining Minimist, and might still be very interested. (Not offering an up-to-date source for git history.)

**Comment by emilbayes** [comment]
@ljharb I can :)

**Comment by ljharb** [comment]
@emilbayes can you push the commits somewhere public so i can grab them?

**Comment by shadowspawn** [comment]
Hi @emilbayes !

I am also willing and able to be a steward of Minimist. I maintain Commander and very familiar with CLI parsing, and have been keeping an eye on Minimist for the last couple of years for interest. I offered an explanation on one issue (not in wayback so second-hand reference): https://github.com/meszaros-lajos-gyorgy/minimist-lite/issues/2

**Comment by ljharb** [comment]
minimist is now covered: https://twitter.com/ljharb/status/1579610392414007299
